eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Arbitrary.hs

```diff
@@ -193,39 +179,208 @@ instance
+    where
+      chooseLineage ::
+        Lens' (PForest PRoot era) (PRoot (GovPurposeId p era)) ->
+        Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId p era)) ->
+        Proposals era ->
+        Seq.Seq (GovActionId (EraCrypto era)) ->
+        Gen (Seq.Seq (GovActionId (EraCrypto era)))
+      chooseLineage lenzR lenzH ps = \case
+        Seq.Empty ->
+          let children = ps ^. pRootsL . lenzR . prChildrenL
+           in if Set.null children
+                then pure Seq.Empty
+                else do
+                  child <- elements $ toList children
+                  chooseLineage lenzR lenzH ps (Seq.Empty Seq.:|> unGovPurposeId child)
+        lineage@(_ Seq.:|> gai) ->
+          let children = ps ^. pHierarchyL . lenzH . pHierarchyNodesL . to (Map.! GovPurposeId gai) . pnChildrenL
+           in if Set.null children
+                then pure lineage
+                else do
+                  child <- elements $ toList children
+                  chooseLineage lenzR lenzH ps (lineage Seq.:|> unGovPurposeId child)
+      consumeHeadAtIndex :: Int -> Seq.Seq (Seq.Seq a) -> (a, Seq.Seq (Seq.Seq a))
+      consumeHeadAtIndex idx ss = (ss `Seq.index` idx `Seq.index` 0, Seq.adjust' (Seq.drop 1) idx ss)
+      sequenceLineages ::
+        Seq.Seq (Seq.Seq (GovActionId (EraCrypto era))) ->
+        Seq.Seq (GovActionId (EraCrypto era)) ->
+        Gen (Seq.Seq (GovActionId (EraCrypto era)))
+      sequenceLineages lineages sequenced = case lineages of
+        Seq.Empty -> pure sequenced
+        _ -> do
+          index <- chooseInt (0, length lineages - 1)
+          let (chosen, adjustedLineages) = consumeHeadAtIndex index lineages
+          sequenceLineages (Seq.filter (not . Seq.null) adjustedLineages) (sequenced Seq.:|> chosen)
+
+data ProposalsNewActions era = ProposalsNewActions (Proposals era) [GovActionState era]
+  deriving (Show, Eq)
+
+instance
+  ( Era era
+  , Arbitrary (PParamsUpdate era)
+  , Arbitrary (PParamsHKD StrictMaybe era)
+  ) =>
+  Arbitrary (ProposalsNewActions era)
+  where
+  arbitrary = do
+    ps <- arbitrary
+    i <- chooseInt (2, 20)
+    gass <- vectorOf i $ genGovActionState =<< genGovAction ps
+    pure $ ProposalsNewActions ps gass
 
 instance
   ( Era era
   , Arbitrary (PParamsUpdate era)
+  , Arbitrary (PParamsHKD StrictMaybe era)
   ) =>
   Arbitrary (Proposals era)
   where
-  arbitrary = fromGovActionStateSeq <$> uniqueIdGovActions
+  arbitrary = genProposals (2, 30)
 
-genGovActionStateFromAction :: Era era => GovAction era -> Gen (GovActionState era)
-genGovActionStateFromAction act =
+genProposals ::
+  forall era.
+  ( Era era
+  , Arbitrary (PParamsHKD StrictMaybe era)
+  ) =>
+  (Int, Int) ->
+  Gen (Proposals era)
+genProposals range = do
+  pgais <- arbitrary
+  i <- chooseInt range
+  go (def & pRootsL .~ fromPrevGovActionIds pgais) i
+  where
+    go :: Proposals era -> Int -> Gen (Proposals era)
+    go ps = \case
+      0 -> pure ps
+      n -> do
```

Comment:
If you supply a negative number here by mistake you'll get an infinite loop
```suggestion
    go ps n
      | n <= 0 = pure ps
      | otherwise = do
```